nixos/modules/services/system/hot-resize.nix

```diff
@@ -0,0 +1,143 @@
+{
+  config,
+  lib,
+  pkgs,
+  ...
+}:
+
+let
+  cfg = config.services.hotResize;
+
+  resize-script = pkgs.writeShellApplication {
+    name = "hot-resize";
+    runtimeInputs = with pkgs; [
+      cloud-utils
+      e2fsprogs
+      parted
+      util-linux
+      btrfs-progs
+      xfsprogs
+      coreutils
+    ];
+
+    text = ''
+      process_device() {
+        local device="$1"
+        local fstype="$2"
+        local mountpoint="$3"
+
+        echo "Processing $device ($fstype) mounted at $mountpoint..."
+
+        REAL_DEVICE=$(readlink -f "$device")
+        DISK=$(lsblk -ndo pkname "$REAL_DEVICE")
+        PART_NUM=$(lsblk -ndo part "$REAL_DEVICE")
+
+        if [ -z "$DISK" ] || [ -z "$PART_NUM" ]; then
+          echo "Failed to determine disk and partition number for $device"
+          return 1
+        fi
+
+        echo "Growing partition /dev/$DISK partition $PART_NUM..."
+        GROWPART_OUTPUT=$(growpart "/dev/$DISK" "$PART_NUM" 2>&1) || {
+          EXIT_CODE=$?
+          if [ $EXIT_CODE -eq 2 ]; then
+            echo "Partition is already at maximum size, continuing with filesystem resize..."
+          else
+            echo "Failed to grow partition: $GROWPART_OUTPUT"
+            return $EXIT_CODE
+          fi
+        }
+
+        echo "Resizing filesystem..."
+        case "$fstype" in
+          ext4)
+            resize2fs "$device"
+            ;;
+          xfs)
+            xfs_growfs "$mountpoint"
+            ;;
+          btrfs)
+            btrfs filesystem resize max "$mountpoint"
+            ;;
+          *)
+            echo "Unsupported filesystem type: $fstype"
+            return 1
+            ;;
+        esac
+
+        echo "Current size:"
+        df -h "$mountpoint"
+        echo "---"
+      }
+
+      process_device "$@"
+    '';
+  };
+
+in
+{
+  options.services.hotResize = {
+    enable = lib.mkEnableOption "the hot resize service for filesystems";
+
+    devices = lib.mkOption {
+      type = lib.types.listOf (
+        lib.types.submodule {
+          options = {
+            device = lib.mkOption {
+              type = lib.types.str;
+              example = "/dev/sda1";
+              description = "Block device path to resize";
+            };
+
+            fsType = lib.mkOption {
+              type = lib.types.enum [
+                "ext4"
+                "xfs"
+                "btrfs"
+              ];
+              default = "ext4";
+              description = "Filesystem type (supported: ext4, xfs, btrfs)";
+            };
+
+            mountPoint = lib.mkOption {
```

Comment:
Maybe a good idea to imitate hardware-configuration.nix, making mount point the name of attributes instead of list

Reply:
Also this module may be more appropriate in tools like disko, rather than a standalone NixOS module

Reply:
The goal was to not use disko that why have create this module.

Reply:
Not sure what your intention is, having a dedicated program is much better than throwing random solutions everywhere, especially when you are trying to do a one shot job with a NixOS module thats gonna be persistent...

Reply:
Disko is the ideal place for something like this, especially since Disko is also written in `bash`.